lab/lablib/better_mouse_events.js

```diff
@@ -0,0 +1,135 @@
+import { fix_mouse_coords } from "../../sketchlib/fix_mouse_coords.js";
+import { prevent_mobile_scroll } from "../../sketchlib/prevent_mobile_scroll.js";
+import { MouseInCanvas, MouseInput, MousePressed } from "./MouseInput.js";
+import { SCREEN_RECT } from "./Rectangle.js";
+/**
+ * @typedef {function(MouseInput):void} MouseCallback
+ */
+
+/**
+ * Class that helps set up canvas events
+ */
+export class MouseHandler {
+  constructor() {
+    /**
+     * @type {HTMLCanvasElement}
+     */
+    this.canvas = undefined;
+  }
+
+  /**
+   * Store the canvas, and prevent mobile scrolling
+   * @param {HTMLCanvasElement} canvas The canvas element
+   */
+  setup(canvas) {
+    this.canvas = canvas;
+    prevent_mobile_scroll(canvas);
+  }
+
+  /**
+   * Create a better mouse pressed event that:
+   * - uses fix_mouse_coords to adjust mouse coords even if the
+   * - ignores events outside the canvas
+   * - events in the canvas will cancel the browser default behavior since this
+   * can interfere with the sketch. Events outside the canvas are passed through
+   * as usual
+   *
+   * Note that this assumes the default canvas dimensions of WIDTH and HEIGHT
+   * @param {any} p the p5.js element
+   * @param {MouseCallback} callback The callback to call when the mouse is pressed on the canvas
+   */
+  mouse_pressed(p, callback) {
+    p.mousePressed = (/** @type {MouseEvent} */ e) => {
+      const mouse = fix_mouse_coords(this.canvas, p.mouseX, p.mouseY);
+      if (!SCREEN_RECT.contains(mouse)) {
+        return;
```

Comment:
I'm feeling a little bit iffy here, maybe it should always call the event and let the caller decide whether to filter out off-canvas events? Maybe there's different variations on the event to select from? e.g. `mouse_pressed_canvas` vs `mouse_pressed`?

Let's see how I use this in practice first though.

Reply:
I wrote this down in my notes